Enclose Logical Expressions with Parentheses

alixaxel opened the pull request:
This PR fixes #232 and #230.

For cases where an expression operator is `&&` or `||`, it surrounds the whole expression with `(...)`, respecting left association. The top-most parentheses is unnecessary but I couldn't figure out how to skip it.

rockwotj commented on src/ast.ts:
```diff
@@ -705,6 +705,11 @@ export function decodeExpression(exp: Exp, outerPrecedence?: number): string {
+
+    if ((innerPrecedence >= outerPrecedence) && ((expOp.op === '&&') || (expOp.op === '||'))) {
+      result = '(' + result + ')';
+    }
```
Can we move this to the `else if (expOp.args.length === 2)` clause?

Also, this comment from above, seems like it's supposed to handle this:

```
// All ops are left associative - so nudge the innerPrecendence
// down on the right hand side to force () for right-associating
// operations.
```

@tomlarkworthy are you familiar with any of this code?

alixaxel replied:
I noticed the comment and wondered about it, but as the tests show, it doesn't seem to be working as the comment describes.

I'll push the other change in a few hours.

alixaxel commented:
@rockwotj I've updated the docs to describe the new output (some of it was already outdated, namely `Null` checks) and I've moved the condition to the the `expOp.args.length === 2` scope (I had originally done this, but ended up moving it further down to cover any unforeseen scenario).

alixaxel commented:
@rockwotj Any update on this? :) I have a +6k LOC generated rules file that needs this fix to be deployed.

rockwotj commented:
Thanks for the ping, this slipped. This has been released as `v0.8.4`.

https://github.com/firebase/bolt/releases/tag/v0.8.4
